File: js/sound.js
async function loadSoundAssets() {
    const poolSize = 4;

    async function loadSound(name) {
        let audio = new Audio("sound/" + name + ".wav");
        audio.loop = false;

        let result;

        if (audio.readyState != 4) {
            console.debug("audio (" + name + ") is not ready. Loading...");
            result = new Promise((resolve, reject) => {
                audio.oncanplaythrough = () => {
                    console.debug("audio (" + name + ") loaded.");
                    resolve(audio);
                    audio.oncanplaythrough = null;
                };

                audio.onerror = () => {
                    console.warn("audio (" + name + ") error.");
                    reject();
                    audio.onerror = null;
                };
            });
        } else {
            console.debug("audio (" + name + ") is ready.");
            result = audio;
        }

        audio.load();

        return result;
    }

    async function makeSoundPool(name) {
        let promises = [];
        for (let i = 0; i < poolSize; i++) {
            promises.push(loadSound(name));
        }

        return {
            pool: await Promise.all(promises),
            play: function () {
                let audio = this.pool.pop();

                if (audio) {
                    audio.loop = false;
                    audio.onended = (event) => {
                        console.debug("audio ended");
                        this.pool.push(audio);
                        audio.onended = undefined;
                    };

                    audio.play();
                } else {
                    console.warn("Out of audio object for '" + name + "'");
                }
            }
        };
    }

    sounds = [
        "bang",
        "posebomb",
        "sac",
        "pick",
        "player_die",
        "oioi",
        "ai",
        "addplayer",
        "victory",
        "draw",
        "clock",
        "time_end"
    ]

    let result = {};
    for (let name of sounds) {
        result[name] = await makeSoundPool(name);
    }

    return result;
}

class SoundManager {
    constructor() {
    }

    async init() {
        this.soundAssets = await loadSoundAssets();
    }

    playSound(name) {
        if (this.soundAssets && this.soundAssets[name]) {
            this.soundAssets[name].play();
        }
    }
}

class MusicManager {
    playlist;
    audio;
    constructor(playlist) {
        this.playlist = playlist;
    }

    start(song) {
        this.stop();
        this.next(song);
    }

    stop() {
        if (this.audio) {
            this.audio.pause();
            this.audio = null;
        }
    }

    next(song) {
        if (!args.includes("-z")) {
            this.stop();

            if (!song) {
                song = Math.floor(Math.random() * this.playlist.length);
            }

            this.audio = new Audio(this.playlist[song]);
            this.audio.volume = 0.7;
            this.audio.loop = true;
            this.audio.play();
        }
    }
}

File: UWP/Sound.cs
using System;
using Microsoft.Xna.Framework.Audio;
using Microsoft.Xna.Framework.Content;
using Microsoft.Xna.Framework.Media;

namespace MrBoom
{
    public class SoundAssets
    {
        public class Sound
        {
            private SoundEffect sound;

            public Sound(SoundEffect sound)
            {
                this.sound = sound;
            }

            public void Play()
            {
                sound.Play();
            }
        }

        public class Music
        {
            private Song song;

            public Music(Song song)
            {
                this.song = song;
            }

            public void Play()
            {
                MediaPlayer.Play(song);
            }
        }

        public Sound Bang;
        public Sound PoseBomb;
        public Sound Sac;
        public Sound Pick;
        public Sound PlayerDie;
        public Sound Oioi;
        public Sound Ai;
        public Sound Addplayer;
        public Sound Victory;
        public Sound Draw;
        public Sound Clock;
        public Sound TimeEnd;
        public Music[] Musics;

        public static SoundAssets Load(ContentManager content)
        {
            Sound loadSound(string name)
            {
                return new Sound(content.Load<SoundEffect>("sound\\" + name));
            }

            Music loadMusic(string name)
            {
                return new Music(content.Load<Song>("music\\" + name));
            }

            return new SoundAssets()
            {
                Bang = loadSound("bang"),
                PoseBomb = loadSound("posebomb"),
                Sac = loadSound("sac"),
                Pick = loadSound("pick"),
                PlayerDie = loadSound("player_die"),
                Oioi = loadSound("oioi"),
                Ai = loadSound("ai"),
                Addplayer = loadSound("addplayer"),
                Victory = loadSound("victory"),
                Draw = loadSound("draw"),
                Clock = loadSound("clock"),
                TimeEnd = loadSound("time_end"),
                Musics = new Music[]
                {
                    loadMusic("anar11"),
                    loadMusic("chipmunk"),
                    loadMusic("chiptune"),
                    loadMusic("deadfeel"),
                    loadMusic("drop"),
                    loadMusic("external"),
                    loadMusic("matkamie"),
                    loadMusic("unreeeal"),
                }
            };
        }
    }

    [Flags]
    public enum Sound
    {
        Bang = 2 << 0,
        PoseBomb = 2 << 1,
        Sac = 2 << 2,
        Pick = 2 << 3,
        PlayerDie = 2 << 4,
        Oioi = 2 << 5,
        Ai = 2 << 6,
        Addplayer = 2 << 7,
        Victory = 2 << 8,
        Draw = 2 << 9,
        Clock = 2 << 10,
        TimeEnd = 2 << 11,
    }
}
